### ISSUE TITLE
String notation and pattern matching: “No interpretation for string [...]”

### ISSUE THREAD

**Issue by lthms** [coq/coq#11237]
#### Description of the problem

I am currently experimenting with string notation, and as far as I can tell, it looks like it does not work with a `match ... with ... end` construction, at least when used as a pattern. I have made a quick research in the currently opened issues and found noting related.

Here is a minimal example that triggers the issue I am facing.

```coq
Inductive bytes := wrap_bytes { unwrap_bytes : list byte }.

Delimit Scope bytes_scope with bytes.
Bind Scope bytes_scope with bytes.
String Notation bytes wrap_bytes unwrap_bytes : bytes_scope.

Open Scope bytes_scope.

Example test_match :=
  match "foo" with
  | "foo" => "bar"
  | "bar" => "foo"
  | x => x
  end.
```

It says there is no interpretation for string "foo", even if `Eval compute in "foo"` works.

Note that the following snippet actually works.

```coq
From Coq Require Import String.

Open Scope string_scope.

Example test_match :=
  match "foo" with
  | "foo" => "bar"
  | "bar" => "foo"
  | x => x
  end.
```

#### Coq Version

```
The Coq Proof Assistant, version 8.10.2 (December 2019)
compiled on Dec 3 2019 14:32:10 with OCaml 4.08.0
```

**Comment by lthms** [comment]
FYI, the newly released `coq.8.11+beta1` version is impacted by this bug as well (which is probably not a surprise, but at least now we know)

**Comment by JasonGross** [comment]
The issue is that `"foo"%bytes` interprets to `wrap_bytes (@cons byte x66 (@cons byte x6f (@cons byte x6f (@nil byte))))`, but `"foo"%string` interprets to `String (Ascii.Ascii false true true false false true true false) (String (Ascii.Ascii true true true true false true true false) (String (Ascii.Ascii true true true true false true true false) EmptyString))`.  If you stick the latter as a pattern in a `match`, statement, it works fine.  If you stick the former as a pattern in a `match` statement, you get the error `Error: The parameters do not bind in patterns; they must be replaced by '_'.` on all of the occurrences of `byte`.

So string notations (and numeral notations) will work as patterns if and only if they recursively use only inductives without parameters (and without non-inductive constructor arguments).

I agree, though, that the error message is quite misleading.

The actual error is presumably a result of `check_allowed_ref_in_pat` in `interp/notation.ml`.

I don't know enough of the infrastructure around this part of the codebase to know how to get a better error message, nor how to, say, convince the notation parsing machinery to drop the parameters when notations are used in match pattern positions.  I suspect @herbelin knows more (or would know which dev would know more).

**Comment by lthms** [comment]
Oh, thanks a lot for this insight!

This means that I could fix my issue if I was using a specialized list rather than the polymorphic one from the standard library?

**Comment by JasonGross** [comment]
> This means that I could fix my issue if I was using a specialized list rather than the polymorphic one from the standard library?

Yes

**Comment by petemud** [comment]
@lthms You can still use polymorphic list like this
```coq
Inductive bytes := Bnil | Bcons : Byte.byte -> bytes -> bytes.

Fixpoint bytes_to_list b :=
  match b with
  | Bnil => nil
  | Bcons x xs => cons x (bytes_to_list xs)
  end.

Fixpoint bytes_from_list l :=
  match l with
  | nil => Bnil
  | cons x xs => Bcons x (bytes_from_list xs)
  end.

String Notation list bytes_from_list bytes_to_list
  (via bytes mapping [[nil] => Bnil, [cons] => Bcons]) : list_scope.

Open Scope list_scope.

Example test_match :=
  match "foo" with
  | "foo" => "bar"
  | "bar" => "foo"
  | x => x
  end.
```

**Comment by herbelin** [comment]
Brought to the attention and fixed with PR #17902.
